File: Adam_FastQ_BaseQual_Baseline/src/main/scala/FastQ_PerBaseQual_Mapper.scala
package main.scala

import org.apache.hadoop.io.Text
import org.bdgenomics.formats.avro.AlignmentRecord

import scala.collection.mutable.ListBuffer

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 05.02.16.
  */
object FastQ_PerBaseQual_Mapper {
  private val OFFSET: Int = 33

  def mapSampleIdentifierWithConvertedInputObject(record: Pair[String, Text]) : Pair[String, AlignmentRecord] = {
    new Pair(record._1, CustomFastqRecordConverter.convertRead(record))
  }

  def flatMap(sampleIdentifier: String, record: AlignmentRecord): TraversableOnce[Pair[Pair[String, Int], Int]] = {
    val resList = new ListBuffer[Pair[Pair[String, Int], Int]]()

    val quality: String = record.getQual
    for(i <- 0 until quality.length) {
      resList.append(new Pair(new Pair(sampleIdentifier, (i+1)), getCorrespoindingIntValue(quality.charAt(i))))
    }
    resList.toTraversable
  }

  def map(pair: Pair[Pair[String, Int], AvgCount]): Pair[Pair[String, Int], Double] = {
    new Pair(pair._1, pair._2.getMeanValue())
  }

  private def getCorrespoindingIntValue(qualVal: Int): Int = {
    (qualVal - OFFSET)
  }

}

File: MapReduce_FastQ_BaseQual_SingleFile_Optimized/src/main/java/baseQual/FastQ_PerBaseQual_Mapper.java
package baseQual;

import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Text;
import org.seqdoop.hadoop_bam.SequencedFragment;
import utils.QualityCountHelperWritable;

import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 21.04.2016
 */
public class FastQ_PerBaseQual_Mapper extends org.apache.hadoop.mapreduce.Mapper<Object, SequencedFragment, IntWritable, QualityCountHelperWritable> {
    private static final int OFFSET = 33;

    private Map<Integer, QualityCountHelperWritable> keyMapping = new HashMap<Integer, QualityCountHelperWritable>();

    @Override
    protected void map(Object key, SequencedFragment value, Context context) throws IOException, InterruptedException {
        final Text quality = value.getQuality();
        for (int i = 0; i < quality.getLength(); i++) {
            int outputKey = (i + 1);
            int qualValue = getCorrespondingIntValue(quality.charAt(i));

            if(!this.keyMapping.containsKey(outputKey)) {
                this.keyMapping.put(outputKey, new QualityCountHelperWritable(qualValue));
            } else {
                this.keyMapping.get(outputKey).addQualityValueToSum(qualValue);
                this.keyMapping.get(outputKey).incrementNumberOfQualityValues();
            }
        }
    }

    @Override
    protected void cleanup(Context context) throws IOException, InterruptedException {
        Iterator it = this.keyMapping.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry pair = (Map.Entry)it.next();
            context.write(new IntWritable((Integer) pair.getKey()), (QualityCountHelperWritable) pair.getValue());
        }
    }

    private int getCorrespondingIntValue(int qualVal) {
        return qualVal - FastQ_PerBaseQual_Mapper.OFFSET;
    }

}